src/Datadog.Trace/Tracer.cs

```diff
@@ -671,7 +671,39 @@ private void InitializeLibLogScopeEventSubscriber(IScopeManager scopeManager, st
 
         private void CurrentDomain_ProcessExit(object sender, EventArgs e)
         {
-            RunShutdownTasks();
+            // This handles a grateful shutdown or a SIGTERM
+            // Multiple delegates can be registered to this event. (eg. IHostApplicationLifetime instances)
+            // So we must be sure we are the last handler of the event to ensure we are flushing all the
+            // possible traces we are generating.
+
+            // For this, we try to get the internal ProcessExit delegate from the event.
+            var processExitDelegate = DelegatesHelper.GetInternalProcessExitDelegate();
+            if (processExitDelegate != null)
+            {
+                // With the internal MulticastDelegate we first create the delegate instance
+                // we want to run at last of the event handler.
+
+                // Sets the final delegate to be executed by the ProcessExit event.
+                var lastDelegate = new EventHandler((s, earg) =>
+                {
+                    RunShutdownTasks();
+                });
+
+                // Try to modify the MulticastDelegate invocation list and set the last delegate.
+                if (!DelegatesHelper.TrySetLastDelegate(processExitDelegate, lastDelegate))
+                {
+                    // This means we are in the last delegate already
+                    // or we were unable to set the last delegate, in
+                    // any case we fallback by running the delegate now.
+                    lastDelegate(sender, e);
```

Comment:
Do we need to call `lastDelegate(sender, e)`? Can we just run `RunShutdownTasks()` like in the `else` fallback case below?

Reply:
done in f861559